providers/fips/self_test_kats.c

```diff
@@ -723,3 +731,122 @@ int SELF_TEST_kats(OSSL_SELF_TEST *st, OSSL_LIB_CTX *libctx)
+};
+static const unsigned char kat_persstr[] = {
+    0x88, 0x6f, 0x54, 0x9a, 0xad, 0x1a, 0xc6, 0x3d, 0x18, 0xcb, 0xcc, 0x66,
+    0x85, 0xda, 0xa2, 0xc2, 0xf7, 0x9e, 0xb0, 0x89, 0x4c, 0xb4, 0xae, 0xf1,
+    0xac, 0x54, 0x4f, 0xce, 0x57, 0xf1, 0x5e, 0x11
+};
+
+// Unpublished function in providers/fips/crypto/rand/rand_lib.c
+EVP_RAND_CTX *RAND_set0_private(OSSL_LIB_CTX *ctx, EVP_RAND_CTX *newrand);
+
+// The default 'private' DRBG of the library context,
+// saved for the duration of KAT testing
+static EVP_RAND_CTX *parent_rand = NULL;
+static EVP_RAND_CTX *saved_rand = NULL;
+
+static int set_kat_drbg(OSSL_LIB_CTX *ctx) {
+    EVP_RAND *rand;
+    EVP_RAND_CTX *kat_rand = NULL;
+    unsigned int strength = 256;
+    OSSL_PARAM drbg_params[3] = {
+        OSSL_PARAM_END, OSSL_PARAM_END, OSSL_PARAM_END
+    };
+
+    // If not NULL, we didn't cleanup from last call: BAD
+    if (parent_rand || saved_rand)
+        goto err;
+
+    rand = EVP_RAND_fetch(ctx, "TEST-RAND", NULL);
+    if (rand == NULL)
+        goto err;
+
+    parent_rand = EVP_RAND_CTX_new(rand, NULL);
+    EVP_RAND_free(rand);
+    if (parent_rand == NULL)
+        goto err;
+
+    drbg_params[0] = OSSL_PARAM_construct_uint(OSSL_RAND_PARAM_STRENGTH, &strength);
+    if (!EVP_RAND_CTX_set_params(parent_rand, drbg_params))
+        goto err;
+
+    rand = EVP_RAND_fetch(ctx, "HASH-DRBG", NULL);
+    if (rand == NULL)
+        goto err;
+
+    kat_rand = EVP_RAND_CTX_new(rand, parent_rand);
+    EVP_RAND_free(rand);
+    if (kat_rand == NULL)
+        goto err;
+
+    drbg_params[0] = OSSL_PARAM_construct_utf8_string("digest", "SHA256", 0);
+    if (!EVP_RAND_CTX_set_params(kat_rand, drbg_params))
+        goto err;
+
+    drbg_params[0] =
+        OSSL_PARAM_construct_octet_string(OSSL_RAND_PARAM_TEST_ENTROPY,
+                                          (void *) kat_entropyin,
+                                          sizeof(kat_entropyin));
+    drbg_params[1] =
+        OSSL_PARAM_construct_octet_string(OSSL_RAND_PARAM_TEST_NONCE,
+                                          (void *) kat_nonce, sizeof(kat_nonce));
+    if (!EVP_RAND_instantiate(parent_rand, strength, 0, NULL, 0, drbg_params))
+        goto err;
+    if (!EVP_RAND_instantiate(kat_rand, strength, 0, kat_persstr, sizeof(kat_persstr), NULL))
+        goto err;
+
+    // Update the library context DRBG
+    saved_rand = RAND_set0_private(ctx, kat_rand);
+    return (saved_rand != NULL);
+err:
+    EVP_RAND_CTX_free(parent_rand);
+    EVP_RAND_CTX_free(kat_rand);
+    return 0;
+}
+
+static void reset_original_drbg(OSSL_LIB_CTX *ctx) {
+    EVP_RAND_CTX *kat_rand = NULL;
+    if (saved_rand == NULL)
+        return;
+
+    // Reset
+    kat_rand = RAND_set0_private(ctx, saved_rand);
+
+    // Cleanup
+    EVP_RAND_uninstantiate(kat_rand);
+    EVP_RAND_verify_zeroization(kat_rand);
+    EVP_RAND_CTX_free(kat_rand);
+    EVP_RAND_CTX_free(parent_rand);
+    parent_rand = NULL;
+    saved_rand = NULL;
+}
+
```

Comment:
We considered the approach with predictable DRBG but considered it quite risky because an implementation error will cause using the predictable DRBG outside of the KAT context.

Reply:
Yes, that is a concern.  I don't see a nice way around this risk.

Reply:
We explicitly pass the random value and use it only in case we are in FIPS self-test mode

Reply:
I've added verification the the TEST-RAND has been replaced after the tests are run.  This should avoid this possibility.